### ISSUE TITLE
[jvm] TObjectDecl code looks a bit wrong

### ISSUE THREAD

**Issue by Simn** [HaxeFoundation/haxe#9802]
I don't have a reproducible example yet, but this code which handles object declaration construction looks like it could break:

```ocaml
					let rec loop fl fl' ok acc = match fl,fl' with
						| ((name,_,_),e) :: fl,(name',jsig) :: fl' ->
							if ok && name = name' then begin
								self#texpr rvalue_any e;
								jm#cast jsig;
								loop fl fl' ok acc
							end else begin
								let load = match (Texpr.skip e).eexpr with
								| TConst _ | TTypeExpr _ | TFunction _ ->
									(fun () -> self#texpr rvalue_any e)
								| _ ->
									let _,load,save = jm#add_local (Printf.sprintf "_hx_tmp_%s" name) (self#vtype e.etype) VarWillInit in
									self#texpr rvalue_any e;
									save();
									load
								in
								loop fl fl' false ((name,load) :: acc)
							end
						| [],[] ->
							acc
						| (_,e) :: fl,[] ->
							self#texpr RVoid e;
							loop fl fl' ok acc
						| [],(_,jsig) :: fl' ->
							jm#load_default_value jsig;
							loop [] fl' ok acc
					in
```

In particular, `jm#load_default_value jsig;` uses the signature of missing fields, but I don't think it's guaranteed that we're actually looking at the field that is truly missing. It could have been covered by an earlier `name <> name'` situation, in which case we're losing synchronization here.

**Comment by Simn** [comment]
Hmm right, I forgot that object declarations are always have their own structural type, so this isn't really an issue:

```haxe
typedef Td = {
	var given:String;
	var ?missing1:Int;
	var ?missing2:String;
}

function main() {
	var obj:Td = {missing1: 1, given: "foo"};
	trace(obj);
}
```

Dump:

```
			[Var obj(3290):Td]
				[ObjectDecl:{ missing1 : Null<Int>, given : java.lang.String }]
					missing1: [Cast:Null<Int>] [Const:Int] 1
					given: [Const:java.lang.String] "foo"
```

Decompiled:

```java
public class Anon0 extends DynamicObject {
    public String given;
    public Integer missing1;

    public Anon0(String given, Integer missing1) {
        super((EmptyConstructor)null);
        this.given = given;
        this.missing1 = missing1;
    }
}

java.lang.Object obj = new Anon0("foo", (Integer)1);
```

For the JVM target it would actually be better if this constructed an instance of `Td` with default values for the missing fields.

This also makes me realize that any `?p:haxe.PosInfos` call generates a full `TAnon` with at least 4 real `tclass_field` instances which doesn't seem ideal. Though that likely only affects unit tests when it comes to compilation performance.

**Comment by RealyUniqueName** [comment]
Maybe it's better to collect declared anon classes in the generator's context? And then upon each `TObjectDecl` check if it fits any of previously declared classes. Such approach could significantly reduce the size of target code generated.

**Comment by Simn** [comment]
It wouldn't affect the size of the target code, but it could affect memory consumption while compiling quite a bit. The only problem is that I still don't know how to identify anon types properly.

**Comment by Simn** [comment]
Well... I just addressed this and then found #2928, so fuck me I guess and glory to reflection.
